### DISCUSSION TITLE
Proper way to remote debugging in NextJS

### DISCUSSION THREAD

**Discussion by davidchingmy** [vercel/next.js#45936] [Help]
Like any other node app, we can enable remote debugging using `NODE_OPTIONS='--inspect=0.0.0.0:9229' next dev`. I have setup this together with docker compose config:

```
  app:
    working_dir: '/app'
    image: 'node:18'
    ports:
      - '3000:3000'
      - '9229:9229'
    restart: unless-stopped
    volumes:
      - '.:/app'
    command: bash -c 'npm install && npm run dev'
```

Confirm that debugger is running:
<img width="568" alt="image" src="https://user-images.githubusercontent.com/103477413/218980596-62fd8288-9b57-446d-bc3e-6eedb2258947.png">

Open inspector browser at chrome://inspect:
<img width="458" alt="image" src="https://user-images.githubusercontent.com/103477413/218981924-c8d8afcc-08ce-4c4e-ae34-585e4d57b04a.png">

No file like _index.tsx or _app.tsx found in Sources tab. Do anyone able to do this?

**Comment by icyJoseph** [comment]
I tried this https://nextjs.org/docs/advanced-features/debugging and I could only see a few files, doing `cmd + P`, I couldn't find the webpack protocol. I wonder if Chrome is blocking it somehow... for security reason perhaps?

**Reply by icyJoseph** [reply]
@davidchingmy I've managed!

![Screenshot 2023-03-29 at 13 41 39](https://user-images.githubusercontent.com/21013447/228524731-4743ea7f-6954-4e41-b065-dc12700c2c0e.png)

I am not sure why but my debugger is not starting where it should, I am still investigating that, but I had to tell Chrome inspect to look for port 9230.

Adding 9230 to `chrome://inspect` (click in `Configure`):

![Screenshot 2023-03-29 at 13 44 02](https://user-images.githubusercontent.com/21013447/228525255-86b38074-75a9-4115-8618-91f0bf8718c0.png)


---

Prints out 9230 as the port to inspect:

<img width="1048" alt="Screenshot 2023-03-29 at 13 42 35" src="https://user-images.githubusercontent.com/21013447/228524918-c261c2b7-e038-4c69-9e46-2af2ae81ee2c.png">

**Reply by icyJoseph** [reply]
There's some more going on at: https://github.com/vercel/next.js/issues/47561

**Reply by taiweituan** [reply]
@icyJoseph, that thread is unrelated to whether you try to do that on a dockerized container. Nonetheless, have you found a way to get it to work?

**Reply by icyJoseph** [reply]
Just make sure all ports are added to the Debugger tool, 9230, is often not added automatically

**Reply by taiweituan** [reply]
Thanks, but still doesn't work for me except for [this comment](https://github.com/vercel/next.js/issues/53127#issuecomment-1835599201). 

Are you still running NextJS inside a container? The debugger only works if I run the NextJS app straight from the terminal (not in a Docker container) for me.

**Reply by icyJoseph** [reply]
Ah right, this issue was about using it through Docker, but then would it not be enough to forward the Debugger ports as well? Probably easier with compose.

**Reply by taiweituan** [reply]
I do use `docker-compose` which has the following setting

```yaml
    ports:
      - 3000:3000
      - 9229:9229
    expose:
      - 9229
```

, but the problem is the `--inspect` needs to specify `0.0.0.0` so we can reach the ports inside the container. 

Currently, the inspector only works by not specifying anything for `--inspect` so the debug instance can increment itself (9229 & 9230). However, if I specify anything for `--inspect` (i.e. `inspect=0.0.0.0`, `inspect=:9229`, or `inspect=0.0.0.0:9229` (which I originally had before upgrading NextJS from older version to 13+ and used to worked), it will throw the

```
/usr/local/bin/node:  must be 0 or in range 1024 to 65535.
```

on app start-up.

**Reply by icyJoseph** [reply]
Isn't --inspect-brk=0.0.0.0 or something like that, that you need? And also do the same for port 9230

**Reply by taiweituan** [reply]
I did, and it doesn't really matter since giving either --inspect or --inspect-brk with IPv4 value will throw 

```
/usr/local/bin/node:  must be 0 or in range 1024 to 65535.
```

**Comment by maurovargascom** [comment]
I'm having the same problem. I can't inspect a Next.js app running on a Docker container using Compose

**Comment by jbccollins** [comment]
Any update on this?

**Comment by dcole2929** [comment]
Same problem. Will add when using the `NODE_OPTIONS='--inspect=0.0.0.0:9229' next dev` I get the following error in console

```
Debugger listening on ws://0.0.0.0:9229/3ed8ecb5-4b4f-44fd-968d-f11acec86c41
For help, see: https://nodejs.org/en/docs/inspector
Starting inspector on 0.0.0.0:9229 failed: address already in use
    the --inspect option was detected, the Next.js router server should be inspected at port 0
```

**Comment by mjanveaux** [comment]
Same problem. Is there a solution for this?

**Comment by chandra2607** [comment]
anyone got a better workaround it?  
I tried various solutions based on some r&d but none of them worked.i am using docker-compose for managing my services.i tried port forwarding with different services but they didn;t work.

@taiweituan @davidchingmy did you got some solution around it ?
